### DISCUSSION TITLE
Get common name from the client certificate

### DISCUSSION THREAD

**Discussion by softshipper** [quarkusio/quarkus#25122] [Q&A] [answered]
Hi all,
I have created a simple REST application that shows how self signed certificate works on Quarkus. The project is hosted on https://github.com/softshipper/playwithmtls.

For example I would like to read the common name from the client certificate and also the expired date.

I tried:
```
import io.quarkus.security.identity.SecurityIdentity
import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType


@Path("/hello")
class ExampleResource(private val identity: SecurityIdentity) {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    fun hello() : String {
        println(identity.principal.name)
        return "Hello"
    }
}
```
but unfortunately I have got a runtime exception:
```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.quarkus.security.identity.SecurityIdentity and qualifiers [@Default]
        - java member: com.example.ExampleResource():identity
        - declared on CLASS bean [types=[com.example.ExampleResource, java.lang.Object], qualifiers=[@Default, @Any], target=com.example.ExampleResource]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:411)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:527)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:263)
        ... 11 more
```
What is missing?

Best regards,
Softshipper

**Comment by softshipper** [comment]
I tried 
```
@Path("/hello")
class ExampleResource(@RequestScoped private val identity: SecurityIdentity) {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    fun hello() : String {
        println(identity.principal.name)
        return "Hello"
    }
} 
```

Unfortunately got the same exception.

**Comment by softshipper** [comment, marked as answer]
I have installed
```
    implementation("io.quarkus:quarkus-smallrye-jwt")
    implementation("io.quarkus:quarkus-smallrye-jwt-build")
```
and afterwards it works.

**Comment by sberyozkin** [comment]
@softshipper, `smallrye-jwt-build` is only needed if JWT tokens have to be generated, so you should be able to drop it.

I also think if MTLS is activated then `SecurityIdentity` should be provided OOB. I'll open an issue to investigate, for example, in this case adding `smallrye-jwt` should not be necessary

**Reply by softshipper** [reply]
@sberyozkin Which extension should I add it to the project? Removing both I have got the exception again. 
Which extension do I need?

**Comment by sberyozkin** [comment]
@softshipper I think it is something wrong with your `IntelliJ` workspace setup, possibly related to #25120 

I've just added a test confirming it works, see #25137. It is a combination of `quarkus-vertx-http` and `quarkus-security` which makes it work

**Reply by softshipper** [reply]
How can I add `quarkus-security` extension to https://github.com/softshipper/playwithmtls? Could you please show it to me.

**Reply by sberyozkin** [reply]
It is here, https://github.com/quarkusio/quarkus/blob/main/extensions/security/runtime/pom.xml

**Reply by softshipper** [reply]
But here https://code.quarkus.io/ `quarkus-security` is not available.

**Reply by sberyozkin** [reply]
I think if you select for example `vertx-http` then it should be included in the list of dependencies. Please open a `code.quarkus.io` issue if it is not the case.
It is not shown on its own because because it is a base dependency which does not offer any specific feature to users.
